File: src/callback_group_learning_cpp/src/deadlock_demo.cpp
/**
 * ===========================================================================
 *  ⑧ 死锁演示与解决 (C++版)
 * ===========================================================================
 *
 * 死锁场景1（经典）：
 *   - 节点有一个服务端和一个定时器
 *   - 定时器回调中调用节点自己的服务
 *   - 如果服务端和定时器在同一个 MutuallyExclusive 回调组中
 *   - 定时器回调持有组锁，等待服务响应
 *   - 服务回调需要获取组锁才能执行
 *   - 结果：死锁！
 *
 * C++ 死锁相关 API：
 *   - 创建服务端: node->create_service<SrvType>(name, callback, group)
 *   - 创建客户端: node->create_client<SrvType>(name, group)
 *   - 异步调用: client->async_send_request(request)
 *   - C++ 中不存在 Python 那样的 rclpy.spin_until_future_complete
 *     但可以用 executor.spin_until_future_complete(future)
 * ===========================================================================
 */

#include <rclcpp/rclcpp.hpp>
#include <example_interfaces/srv/trigger.hpp>
#include <std_msgs/msg/string.hpp>
#include <chrono>
#include <thread>
#include <sstream>

using namespace std::chrono_literals;

// ================================================================
// 死锁版本 —— 演示问题
// ================================================================

class DeadlockNode : public rclcpp::Node
{
public:
    DeadlockNode() : Node("deadlock_node")
    {
        // ================================================================
        // 错误设计：服务端和定时器在同一个 MutuallyExclusive 回调组
        // 这会导致定时器回调中调用服务时产生死锁！
        // ================================================================
        bad_group_ = this->create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);

        // 服务端 —— 在 bad_group_ 中
        srv_ = this->create_service<example_interfaces::srv::Trigger>(
            "/deadlock_trigger",
            std::bind(&DeadlockNode::service_callback, this,
                      std::placeholders::_1, std::placeholders::_2),
            rclcpp::ServicesQoS(), bad_group_);

        // 客户端 —— 用于调用服务
        client_ = this->create_client<example_interfaces::srv::Trigger>(
            "/deadlock_trigger",
            rclcpp::ServicesQoS(), bad_group_);

        // 定时器 —— 也在 bad_group_ 中
        timer_ = this->create_wall_timer(
            2s,
            std::bind(&DeadlockNode::timer_callback, this),
            bad_group_);

        RCLCPP_WARN(this->get_logger(), "================================================================");
        RCLCPP_WARN(this->get_logger(), "死锁演示节点启动 (C++)");
        RCLCPP_WARN(this->get_logger(), "  服务端和定时器在同一 MutuallyExclusive 回调组");
        RCLCPP_WARN(this->get_logger(), "  定时器回调中调用自己的服务 → 会死锁！");
        RCLCPP_WARN(this->get_logger(), "================================================================");
    }

private:
    /**
     * 服务回调 —— 需要 bad_group_ 的锁才能执行
     */
    void service_callback(
        const std::shared_ptr<example_interfaces::srv::Trigger::Request> /*request*/,
        std::shared_ptr<example_interfaces::srv::Trigger::Response> response)
    {
        RCLCPP_INFO(this->get_logger(), "[服务] 收到请求，处理中...");
        std::this_thread::sleep_for(500ms);
        response->success = true;
        response->message = "服务处理完成";
    }

    /**
     * 定时器回调 —— 持有 bad_group_ 的锁
     */
    void timer_callback()
    {
        RCLCPP_INFO(this->get_logger(), "[定时器] 调用自己的服务...");

        // ============================================================
        // 死锁发生点！
        // 1. timer_callback 获取了 bad_group_ 的互斥锁
        // 2. 调用服务需要服务回调执行
        // 3. 服务回调也在 bad_group_ 中，需要获取同一个互斥锁
        // 4. 但互斥锁已被 timer_callback 持有
        // 5. 死锁！
        // ============================================================
        auto request = std::make_shared<example_interfaces::srv::Trigger::Request>();
        auto future = client_->async_send_request(request);
        RCLCPP_INFO(this->get_logger(), "[定时器] 服务调用已发送，等待响应...");

        // 在 MutuallyExclusive 回调组中，服务回调永远不会被执行
        // 因为组锁被定时器回调持有
    }

    rclcpp::CallbackGroup::SharedPtr bad_group_;
    rclcpp::Service<example_interfaces::srv::Trigger>::SharedPtr srv_;
    rclcpp::Client<example_interfaces::srv::Trigger>::SharedPtr client_;
    rclcpp::TimerBase::SharedPtr timer_;
};


// ================================================================
// 正确版本 —— 解决死锁
// ================================================================

class FixedNode : public rclcpp::Node
{
public:
    FixedNode() : Node("fixed_node"), call_count_(0)
    {
        // ================================================================
        // 正确设计：服务端和定时器使用不同的回调组
        // ================================================================
        timer_group_ = this->create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);
        service_group_ = this->create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);
        sub_group_ = this->create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);

        // 服务端 —— 在 service_group_ 中
        srv_ = this->create_service<example_interfaces::srv::Trigger>(
            "/fixed_trigger",
            std::bind(&FixedNode::service_callback, this,
                      std::placeholders::_1, std::placeholders::_2),
            rclcpp::ServicesQoS(), service_group_);

        // 客户端 —— 调用服务
        client_ = this->create_client<example_interfaces::srv::Trigger>(
            "/fixed_trigger",
            rclcpp::ServicesQoS(), service_group_);

        // 定时器 —— 在 timer_group_ 中
        timer_ = this->create_wall_timer(
            2s,
            std::bind(&FixedNode::timer_callback, this),
            timer_group_);

        // 订阅 —— 在 sub_group_ 中
        rclcpp::SubscriptionOptions sub_opts;
        sub_opts.callback_group = sub_group_;
        sub_ = this->create_subscription<std_msgs::msg::String>(
            "/status", 10,
            std::bind(&FixedNode::status_callback, this, std::placeholders::_1),
            sub_opts);

        RCLCPP_INFO(this->get_logger(), "================================================================");
        RCLCPP_INFO(this->get_logger(), "正确设计节点启动 (C++)");
        RCLCPP_INFO(this->get_logger(), "  服务端 → service_group (独立)");
        RCLCPP_INFO(this->get_logger(), "  定时器 → timer_group (独立)");
        RCLCPP_INFO(this->get_logger(), "  订阅   → sub_group (独立)");
        RCLCPP_INFO(this->get_logger(), "  使用 MultiThreadedExecutor → 不会死锁");
        RCLCPP_INFO(this->get_logger(), "================================================================");
    }

private:
    void service_callback(
        const std::shared_ptr<example_interfaces::srv::Trigger::Request> /*request*/,
        std::shared_ptr<example_interfaces::srv::Trigger::Response> response)
    {
        RCLCPP_INFO(this->get_logger(), "[服务] 收到请求，处理中...");
        std::this_thread::sleep_for(500ms);
        call_count_++;
        response->success = true;
        response->message = "服务处理完成，第 " + std::to_string(call_count_) + " 次调用";
    }

    void timer_callback()
    {
        call_count_++;
        RCLCPP_INFO(this->get_logger(),
            "[定时器] 调用服务 (第%d次)...", call_count_);

        // 使用 async_send_request 异步调用
        // 服务端在不同的回调组中，可以并发执行
        auto request = std::make_shared<example_interfaces::srv::Trigger::Request>();
        auto future = client_->async_send_request(request);

        // 注意：不要在这里同步等待 future 的结果！
        // 如果需要结果，使用 spin_until_future_complete 或回调
        RCLCPP_INFO(this->get_logger(), "[定时器] 服务调用已发送");
    }

    void status_callback(const std_msgs::msg::String::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "[订阅] 状态: %s", msg->data.c_str());
    }

    rclcpp::CallbackGroup::SharedPtr timer_group_;
    rclcpp::CallbackGroup::SharedPtr service_group_;
    rclcpp::CallbackGroup::SharedPtr sub_group_;
    rclcpp::Service<example_interfaces::srv::Trigger>::SharedPtr srv_;
    rclcpp::Client<example_interfaces::srv::Trigger>::SharedPtr client_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
    int call_count_;
};


// ================================================================
// 最佳实践版本 —— 避免回调中调用自己的服务
// ================================================================

class BestPracticeNode : public rclcpp::Node
{
public:
    BestPracticeNode() : Node("best_practice_node")
    {
        // 只需要一个回调组
        group_ = this->create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);

        timer_ = this->create_wall_timer(
            2s,
            std::bind(&BestPracticeNode::timer_callback, this),
            group_);

        RCLCPP_INFO(this->get_logger(), "================================================================");
        RCLCPP_INFO(this->get_logger(), "最佳实践节点启动 (C++)");
        RCLCPP_INFO(this->get_logger(), "  不在回调中调用自己的服务");
        RCLCPP_INFO(this->get_logger(), "  而是直接调用内部方法");
        RCLCPP_INFO(this->get_logger(), "================================================================");
    }

private:
    /**
     * 核心逻辑封装在普通方法中（而非服务回调中）
     * ✅ 最佳实践：直接调用方法，避免服务调用的开销和死锁风险
     */
    std::pair<bool, std::string> do_something()
    {
        RCLCPP_INFO(this->get_logger(), "[核心逻辑] 执行处理...");
        std::this_thread::sleep_for(500ms);
        RCLCPP_INFO(this->get_logger(), "[核心逻辑] 处理完成");
        return {true, "处理成功"};
    }

    void timer_callback()
    {
        auto [success, message] = do_something();
        RCLCPP_INFO(this->get_logger(),
            "[定时器] 结果: success=%s, msg=\"%s\"",
            success ? "true" : "false", message.c_str());
    }

    rclcpp::CallbackGroup::SharedPtr group_;
    rclcpp::TimerBase::SharedPtr timer_;
};


// ================================================================
// 主函数 —— 根据命令行参数选择模式
// ================================================================

void print_usage(const char* program)
{
    std::cout << "\n用法: " << program << " [模式]\n"
              << "  1 - 死锁演示（MultiThreadedExecutor + 同组服务调用）\n"
              << "  2 - 正确版本（MultiThreadedExecutor + 独立回调组）\n"
              << "  3 - 最佳实践（直接方法调用，默认）\n" << std::endl;
}

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);

    int mode = 3;  // 默认模式3
    if (argc > 1) {
        try {
            mode = std::stoi(argv[1]);
        } catch (...) {
            print_usage(argv[0]);
            rclcpp::shutdown();
            return 1;
        }
    }

    switch (mode) {
        case 1: {
            // 死锁演示
            auto node = std::make_shared<DeadlockNode>();
            rclcpp::executors::MultiThreadedExecutor executor(
                rclcpp::ExecutorOptions(), 4);
            executor.add_node(node);
            try { executor.spin(); } catch (...) {}
            break;
        }
        case 2: {
            // 正确版本
            auto node = std::make_shared<FixedNode>();
            rclcpp::executors::MultiThreadedExecutor executor(
                rclcpp::ExecutorOptions(), 4);
            executor.add_node(node);
            try { executor.spin(); } catch (...) {}
            break;
        }
        default: {
            // 最佳实践
            auto node = std::make_shared<BestPracticeNode>();
            rclcpp::spin(node);
            break;
        }
    }

    rclcpp::shutdown();
    return 0;
}

File: src/callback_group_learning/callback_group_learning/deadlock_demo.py
#!/usr/bin/env python3
"""
=============================================================================
 ⑧ 死锁演示与解决 —— 回调组常见陷阱
=============================================================================

核心知识点：
  - 死锁：两个或多个回调互相等待，导致都无法继续执行
  - ROS2 中常见的3种死锁模式：
    1. 同一 MutuallyExclusive 回调组内的服务调用死锁
    2. 跨节点服务调用 + MutualExclusiveCallbackGroup 死锁
    3. 单线程执行器中长时间阻塞回调

死锁场景1（经典）：
  - 节点A有一个服务端和一个定时器
  - 定时器回调中调用节点A自己的服务
  - 如果服务端和定时器在同一个 MutuallyExclusive 回调组中
  - 定时器回调持有组锁，等待服务响应
  - 服务回调需要获取组锁才能执行
  - 结果：死锁！

解决方案：
  1. 将服务端和调用方放在不同的回调组
  2. 使用 ReentrantCallbackGroup
  3. 避免在回调中同步调用自己的服务（改用直接函数调用）

死锁场景2（跨节点）：
  - 节点A在回调中调用节点B的服务
  - 节点B在回调中调用节点A的服务
  - 如果双方都在 MutuallyExclusive 回调组中
  - 单线程执行器下必定死锁

本示例：
  - 演示死锁场景1（同节点服务调用）
  - 提供正确的解决方案
=============================================================================
"""

import time
import rclpy
from rclpy.node import Node
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup, ReentrantCallbackGroup
from rclpy.executors import SingleThreadedExecutor, MultiThreadedExecutor
from std_msgs.msg import String
from example_interfaces.srv import Trigger


# ================================================================
# 死锁版本 —— 演示问题
# ================================================================

class DeadlockNode(Node):
    """死锁演示节点 —— 服务端和调用方在同一回调组"""

    def __init__(self):
        super().__init__('deadlock_node')

        # ============================================================
        # 错误设计：服务端和定时器在同一个 MutuallyExclusive 回调组
        # 这会导致定时器回调中调用服务时产生死锁！
        # ============================================================
        self.bad_group = MutuallyExclusiveCallbackGroup()

        # 服务端 —— 在 bad_group 中
        self.srv = self.create_service(
            Trigger, '/deadlock_trigger', self.service_callback,
            callback_group=self.bad_group,  # 与定时器同组
        )

        # 定时器 —— 也在 bad_group 中
        self.timer = self.create_timer(
            2.0, self.timer_callback,
            callback_group=self.bad_group,  # 与服务端同组
        )

        self.get_logger().warn('=' * 60)
        self.get_logger().warn('死锁演示节点启动')
        self.get_logger().warn('  服务端和定时器在同一 MutuallyExclusive 回调组')
        self.get_logger().warn('  定时器回调中调用自己的服务 → 会死锁！')
        self.get_logger().warn('=' * 60)

    def service_callback(self, request, response):
        """服务回调 —— 需要 bad_group 的锁才能执行"""
        self.get_logger().info('[服务] 收到请求，处理中...')
        time.sleep(0.5)
        response.success = True
        response.message = '服务处理完成'
        return response

    def timer_callback(self):
        """定时器回调 —— 持有 bad_group 的锁"""
        self.get_logger().info('[定时器] 调用自己的服务...')

        # ============================================================
        # 死锁发生点！
        # 1. timer_callback 获取了 bad_group 的互斥锁
        # 2. 调用服务需要服务回调执行
        # 3. 服务回调也在 bad_group 中，需要获取同一个互斥锁
        # 4. 但互斥锁已被 timer_callback 持有
        # 5. 死锁！
        # ============================================================

        # 使用 call_async 而非 call，避免同步阻塞
        # 但在同一个 MutuallyExclusive 回调组中，
        # 即使是 call_async，服务回调也无法执行（组锁被持有）
        future = self.srv.call_async(Trigger.Request())
        self.get_logger().info('[定时器] 服务调用已发送，等待响应...')

        # 注意：在 SingleThreadedExecutor 中，
        # 服务回调永远不会被执行（因为组锁被持有）
        # 在 MultiThreadedExecutor 中也是一样！
        # 因为 MutuallyExclusive 回调组的锁被定时器回调持有


# ================================================================
# 正确版本 —— 解决死锁
# ================================================================

class FixedNode(Node):
    """正确设计的节点 —— 服务端和调用方在不同回调组"""

    def __init__(self):
        super().__init__('fixed_node')

        # ============================================================
        # 正确设计1：服务端和定时器使用不同的回调组
        # ============================================================
        self.timer_group = MutuallyExclusiveCallbackGroup()
        self.service_group = MutuallyExclusiveCallbackGroup()

        # 服务端 —— 在 service_group 中
        self.srv = self.create_service(
            Trigger, '/fixed_trigger', self.service_callback,
            callback_group=self.service_group,  # 独立回调组
        )

        # 定时器 —— 在 timer_group 中
        self.timer = self.create_timer(
            2.0, self.timer_callback,
            callback_group=self.timer_group,  # 另一个回调组
        )

        # 订阅示例 —— 也在独立回调组
        self.sub_group = MutuallyExclusiveCallbackGroup()
        self.sub = self.create_subscription(
            String, '/status', self.status_callback, 10,
            callback_group=self.sub_group,
        )

        self.call_count = 0
        self.get_logger().info('=' * 60)
        self.get_logger().info('正确设计节点启动')
        self.get_logger().info('  服务端 → service_group (独立)')
        self.get_logger().info('  定时器 → timer_group (独立)')
        self.get_logger().info('  订阅   → sub_group (独立)')
        self.get_logger().info('  使用 MultiThreadedExecutor → 不会死锁')
        self.get_logger().info('=' * 60)

    def service_callback(self, request, response):
        """服务回调 —— 在 service_group 中，可以与定时器并发"""
        self.get_logger().info('[服务] 收到请求，处理中...')
        time.sleep(0.5)
        response.success = True
        response.message = f'服务处理完成，第 {self.call_count} 次调用'
        return response

    def timer_callback(self):
        """定时器回调 —— 在 timer_group 中"""
        self.call_count += 1
        self.get_logger().info(f'[定时器] 调用服务 (第{self.call_count}次)...')

        # 使用 call_async 异步调用
        # 服务端在不同的回调组中，可以并发执行
        future = self.srv.call_async(Trigger.Request())

        # 添加完成回调（不要在此阻塞等待结果）
        def on_done(fut):
            try:
                result = fut.result()
                self.get_logger().info(
                    f'[定时器] 服务响应: success={result.success}, '
                    f'msg="{result.message}"'
                )
            except Exception as e:
                self.get_logger().error(f'[定时器] 服务调用失败: {e}')

        future.add_done_callback(on_done)

    def status_callback(self, msg: String):
        """状态订阅回调 —— 在 sub_group 中"""
        self.get_logger().info(f'[订阅] 状态: {msg.data}')


# ================================================================
# 最佳实践版本 —— 避免回调中调用服务
# ================================================================

class BestPracticeNode(Node):
    """最佳实践 —— 避免在回调中调用自己的服务"""

    def __init__(self):
        super().__init__('best_practice_node')

        # 只需要一个回调组
        self.group = MutuallyExclusiveCallbackGroup()

        self.timer = self.create_timer(
            2.0, self.timer_callback,
            callback_group=self.group,
        )

        # 核心逻辑封装在普通方法中
        self.get_logger().info('=' * 60)
        self.get_logger().info('最佳实践节点启动')
        self.get_logger().info('  不在回调中调用自己的服务')
        self.get_logger().info('  而是直接调用内部方法')
        self.get_logger().info('=' * 60)

    def do_something(self):
        """核心逻辑封装在普通方法中（而非服务回调中）"""
        self.get_logger().info('[核心逻辑] 执行处理...')
        time.sleep(0.5)
        self.get_logger().info('[核心逻辑] 处理完成')
        return True, '处理成功'

    def timer_callback(self):
        """定时器回调 —— 直接调用内部方法，不走服务"""
        # ✅ 最佳实践：直接调用方法，避免服务调用的开销和死锁风险
        success, message = self.do_something()
        self.get_logger().info(f'[定时器] 结果: success={success}, msg="{message}"')


def main(args=None):
    rclpy.init(args=args)

    print('\n' + '=' * 60)
    print('选择演示模式:')
    print('  1 - 死锁演示（SingleThreadedExecutor）')
    print('  2 - 正确版本（MultiThreadedExecutor + 独立回调组）')
    print('  3 - 最佳实践（直接方法调用）')
    print('=' * 60)

    import sys
    mode = sys.argv[1] if len(sys.argv) > 1 else '3'

    if mode == '1':
        # ============================================================
        # 死锁演示
        # ============================================================
        node = DeadlockNode()
        executor = MultiThreadedExecutor(num_threads=4)
        # 即使使用 MultiThreadedExecutor，
        # 同一 MutuallyExclusive 回调组中的服务回调也无法执行
        # 因为定时器回调持有组锁
        try:
            rclpy.spin(node, executor=executor)
        except KeyboardInterrupt:
            pass
        finally:
            node.destroy_node()

    elif mode == '2':
        # ============================================================
        # 正确版本：不同回调组 + MultiThreadedExecutor
        # ============================================================
        node = FixedNode()
        executor = MultiThreadedExecutor(num_threads=4)
        try:
            rclpy.spin(node, executor=executor)
        except KeyboardInterrupt:
            pass
        finally:
            node.destroy_node()

    else:
        # ============================================================
        # 最佳实践：避免回调中调用自己的服务
        # ============================================================
        node = BestPracticeNode()
        try:
            rclpy.spin(node)
        except KeyboardInterrupt:
            pass
        finally:
            node.destroy_node()

    rclpy.shutdown()


if __name__ == '__main__':
    main()
